File: ui/internal/accountwizard/fields.go
// SPDX-FileCopyrightText: 2026 Vladislav Janeček
// SPDX-License-Identifier: GPL-3.0-or-later

// Package accountwizard is the "Add Account" dialog: identity → discovery →
// server settings → connection test → account.add. The backend discovers,
// tests and validates; this package only checks field syntax, fills port
// defaults and shows results.
package accountwizard

import (
	"net/mail"
	"strings"

	"github.com/schotek/malachi/backend/pkg/api"
)

// Endpoint selects the IMAP or SMTP side of the settings.
type Endpoint int

const (
	EndpointIMAP Endpoint = iota
	EndpointSMTP
)

// Identity is what the first page collects.
type Identity struct {
	DisplayName string
	Email       string
	Password    string
}

// ServerFields mirrors the rows of one endpoint on the Servers page.
type ServerFields struct {
	Host     string
	Port     int
	Security api.Security
	Username string
}

// securityChoices is the order of the "Security" StringList in
// account_wizard.blp.
var securityChoices = []api.Security{api.SecurityTLS, api.SecuritySTARTTLS, api.SecurityNone}

// DefaultPort is the conventional port for a security mode.
func DefaultPort(kind Endpoint, sec api.Security) int {
	switch kind {
	case EndpointIMAP:
		if sec == api.SecurityTLS {
			return 993
		}
		return 143
	default:
		if sec == api.SecurityTLS {
			return 465
		}
		return 587
	}
}

// PortForSecurityChange keeps a custom port and only swaps the default of
// the previous mode for the default of the new one.
func PortForSecurityChange(kind Endpoint, port int, from, to api.Security) int {
	if from == to {
		return port
	}
	if port == 0 || port == DefaultPort(kind, from) {
		return DefaultPort(kind, to)
	}
	return port
}

func indexOfSecurity(s api.Security) uint {
	for i, c := range securityChoices {
		if c == s {
			return uint(i)
		}
	}
	return 0
}

func securityAt(i uint) api.Security {
	if i < uint(len(securityChoices)) {
		return securityChoices[i]
	}
	return api.SecurityTLS
}

// ValidateEmail trims and accepts only a bare address: no display name, no
// comments. The backend validates authoritatively; this is immediate
// feedback.
func ValidateEmail(s string) (string, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return "", false
	}
	a, err := mail.ParseAddress(s)
	if err != nil || a.Name != "" || a.Address != s {
		return "", false
	}
	return s, true
}

// Domain is the lower-cased part after the last '@', or "".
func Domain(email string) string {
	i := strings.LastIndexByte(email, '@')
	if i < 0 || i == len(email)-1 {
		return ""
	}
	return strings.ToLower(email[i+1:])
}

// SuggestAccountName is the default account name: the address's domain.
func SuggestAccountName(email string) string {
	if d := Domain(email); d != "" {
		return d
	}
	return email
}

// GuessConfig is the fallback when discovery finds nothing.
func GuessConfig(email string) api.AccountConfig {
	d := Domain(email)
	return api.AccountConfig{
		Email: email,
		IMAP:  api.ServerConfig{Host: "imap." + d, Port: 993, Security: api.SecurityTLS, Username: email, AuthMethod: api.AuthPassword},
		SMTP:  api.ServerConfig{Host: "smtp." + d, Port: 587, Security: api.SecuritySTARTTLS, Username: email, AuthMethod: api.AuthPassword},
	}
}

// MergeIdentity copies what the identity page knows into a discovered or
// guessed configuration.
func MergeIdentity(cfg api.AccountConfig, id Identity) api.AccountConfig {
	cfg.Email = id.Email
	cfg.DisplayName = strings.TrimSpace(id.DisplayName)
	if strings.TrimSpace(cfg.Name) == "" {
		cfg.Name = SuggestAccountName(id.Email)
	}
	for _, sc := range []*api.ServerConfig{&cfg.IMAP, &cfg.SMTP} {
		if strings.TrimSpace(sc.Username) == "" {
			sc.Username = id.Email
		}
		sc.AuthMethod = api.AuthPassword
	}
	return cfg
}

// IdentityProblems flags the identity fields that cannot be sent.
type IdentityProblems struct{ Email, Password bool }

// Any reports whether anything is wrong.
func (p IdentityProblems) Any() bool { return p.Email || p.Password }

// ValidateIdentity checks address syntax and a non-empty password.
func ValidateIdentity(id Identity) IdentityProblems {
	_, ok := ValidateEmail(id.Email)
	return IdentityProblems{Email: !ok, Password: id.Password == ""}
}

// ServerProblems flags the server rows that cannot be sent.
type ServerProblems struct {
	IMAPHost, IMAPUser, SMTPHost, SMTPUser bool
}

// Any reports whether anything is wrong.
func (p ServerProblems) Any() bool { return p.IMAPHost || p.IMAPUser || p.SMTPHost || p.SMTPUser }

// ValidateServers flags empty hosts and user names. Ports are bounded by
// the spin rows; everything else is the backend's call.
func ValidateServers(imap, smtp ServerFields) ServerProblems {
	return ServerProblems{
		IMAPHost: strings.TrimSpace(imap.Host) == "",
		IMAPUser: strings.TrimSpace(imap.Username) == "",
		SMTPHost: strings.TrimSpace(smtp.Host) == "",
		SMTPUser: strings.TrimSpace(smtp.Username) == "",
	}
}

// BuildConfig assembles the wire configuration from the rows. It is the
// one place that sets the authentication method.
func BuildConfig(id Identity, name string, imap, smtp ServerFields) api.AccountConfig {
	email := strings.TrimSpace(id.Email)
	name = strings.TrimSpace(name)
	if name == "" {
		name = SuggestAccountName(email)
	}
	return api.AccountConfig{
		Name:        name,
		Email:       email,
		DisplayName: strings.TrimSpace(id.DisplayName),
		IMAP:        serverConfig(imap),
		SMTP:        serverConfig(smtp),
	}
}

func serverConfig(f ServerFields) api.ServerConfig {
	return api.ServerConfig{
		Host:       strings.TrimSpace(f.Host),
		Port:       f.Port,
		Security:   f.Security,
		Username:   strings.TrimSpace(f.Username),
		AuthMethod: api.AuthPassword,
	}
}

func credentialsFor(id Identity) api.Credentials {
	return api.Credentials{Password: id.Password}
}

File: macos/Sources/MalachiCore/Wizard/Fields.swift
// SPDX-FileCopyrightText: 2026 Vladislav Janeček
// SPDX-License-Identifier: GPL-3.0-or-later

// ui/internal/accountwizard/fields.go: the field logic of the account
// wizard. The backend discovers, tests and validates; this only checks
// field syntax, fills port defaults and assembles the configuration.

import Foundation

/// accountwizard.Endpoint: the IMAP or SMTP side of the settings.
public enum Endpoint: Sendable, Hashable, CaseIterable {
    case imap
    case smtp
}

/// accountwizard.Identity: what the first page collects.
public struct Identity: Sendable, Equatable {
    public var displayName: String
    public var email: String
    public var password: String

    public init(displayName: String = "", email: String = "", password: String = "") {
        self.displayName = displayName
        self.email = email
        self.password = password
    }
}

/// accountwizard.ServerFields: the rows of one endpoint on the Servers page,
/// and the certificate pinned to it (not a row: the wizard sets it from the
/// pin trusted for this host and port, `CertTrust.keepPin`). "" is no pin.
public struct ServerFields: Sendable, Equatable {
    public var host: String
    public var port: Int
    public var security: Security
    public var username: String
    public var certificateSha256: String

    public init(host: String = "", port: Int = 0, security: Security = .tls, username: String = "", certificateSha256: String = "") {
        self.host = host
        self.port = port
        self.security = security
        self.username = username
        self.certificateSha256 = certificateSha256
    }
}

/// accountwizard.securityChoices: the order of the "Security" choices in
/// account_wizard.blp.
public let securityChoices: [Security] = [.tls, .starttls, .none]

/// accountwizard.DefaultPort: the conventional port for a security mode.
public func defaultPort(_ kind: Endpoint, _ security: Security) -> Int {
    switch kind {
    case .imap:
        return security == .tls ? 993 : 143
    case .smtp:
        return security == .tls ? 465 : 587
    }
}

/// accountwizard.PortForSecurityChange: keeps a custom port and only swaps
/// the default of the previous mode for the default of the new one.
public func portForSecurityChange(_ kind: Endpoint, port: Int, from: Security, to: Security) -> Int {
    if from == to {
        return port
    }
    if port == 0 || port == defaultPort(kind, from) {
        return defaultPort(kind, to)
    }
    return port
}

/// accountwizard.indexOfSecurity: the row of a mode in `securityChoices`,
/// 0 for an unknown one.
public func indexOfSecurity(_ security: Security) -> Int {
    securityChoices.firstIndex(of: security) ?? 0
}

/// accountwizard.securityAt: the mode of a row, TLS out of range.
public func securityAt(_ index: Int) -> Security {
    if index >= 0, index < securityChoices.count {
        return securityChoices[index]
    }
    return .tls
}

/// accountwizard.ValidateEmail: trims and accepts only a bare address: no
/// display name, no comments. The backend validates authoritatively; this
/// is immediate feedback. nil when refused.
public func validateEmail(_ s: String) -> String? {
    let trimmed = s.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.isEmpty {
        return nil
    }
    guard let a = AddressList.parseAddress(trimmed), a.name == nil, a.address == trimmed else {
        return nil
    }
    return trimmed
}

/// accountwizard.Domain: the lower-cased part after the last '@', or "".
public func domain(_ email: String) -> String {
    let utf8 = email.utf8
    guard let i = utf8.lastIndex(of: UInt8(ascii: "@")) else { return "" }
    let after = utf8.index(after: i)
    if after == utf8.endIndex {
        return ""
    }
    return String(decoding: utf8[after...], as: UTF8.self).lowercased()
}

/// accountwizard.SuggestAccountName: the default account name: the
/// address's domain.
public func suggestAccountName(_ email: String) -> String {
    let d = domain(email)
    return d.isEmpty ? email : d
}

/// accountwizard.GuessConfig: the fallback when discovery finds nothing.
public func guessConfig(_ email: String) -> AccountConfig {
    let (imap, smtp) = guessServers(email)
    return AccountConfig(name: "", email: email, kind: .imap, imap: imap, smtp: smtp)
}

private func guessServers(_ email: String) -> (imap: ServerConfig, smtp: ServerConfig) {
    let d = domain(email)
    return (
        ServerConfig(host: "imap." + d, port: 993, security: .tls, username: email, authMethod: .password),
        ServerConfig(host: "smtp." + d, port: 587, security: .starttls, username: email, authMethod: .password)
    )
}

/// accountwizard.MergeIdentity: copies what the identity page knows into
/// a discovered or guessed IMAP configuration. Missing endpoints are
/// filled from the guess so the Servers page always has both.
public func mergeIdentity(_ config: AccountConfig, _ id: Identity) -> AccountConfig {
    var cfg = config
    cfg.email = id.email
    cfg.displayName = nonEmpty(id.displayName.trimmingCharacters(in: .whitespacesAndNewlines))
    if cfg.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        cfg.name = suggestAccountName(id.email)
    }
    let guess = guessServers(id.email)
    cfg.kind = .imap
    cfg.graph = nil
    var imap = cfg.imap ?? guess.imap
    var smtp = cfg.smtp ?? guess.smtp
    if imap.username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        imap.username = id.email
    }
    imap.authMethod = .password
    if smtp.username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        smtp.username = id.email
    }
    smtp.authMethod = .password
    cfg.imap = imap
    cfg.smtp = smtp
    return cfg
}

/// accountwizard.IdentityProblems: the identity fields that cannot be sent.
public struct IdentityProblems: Sendable, Equatable {
    public var email: Bool
    public var password: Bool

    public init(email: Bool = false, password: Bool = false) {
        self.email = email
        self.password = password
    }

    /// Whether anything is wrong.
    public var any: Bool { email || password }
}

/// accountwizard.ValidateIdentity: checks address syntax and, when
/// required, a non-empty password (editing an account may keep the
/// stored one).
public func validateIdentity(_ id: Identity, passwordRequired: Bool) -> IdentityProblems {
    IdentityProblems(email: validateEmail(id.email) == nil, password: passwordRequired && id.password.isEmpty)
}

/// accountwizard.ServerProblems: the server rows that cannot be sent.
public struct ServerProblems: Sendable, Equatable {
    public var imapHost: Bool
    public var imapUser: Bool
    public var smtpHost: Bool
    public var smtpUser: Bool

    public init(imapHost: Bool = false, imapUser: Bool = false, smtpHost: Bool = false, smtpUser: Bool = false) {
        self.imapHost = imapHost
        self.imapUser = imapUser
        self.smtpHost = smtpHost
        self.smtpUser = smtpUser
    }

    /// Whether anything is wrong.
    public var any: Bool { imapHost || imapUser || smtpHost || smtpUser }
}

/// accountwizard.ValidateServers: flags empty hosts and user names. Ports
/// are bounded by the steppers; everything else is the backend's call.
public func validateServers(imap: ServerFields, smtp: ServerFields) -> ServerProblems {
    ServerProblems(
        imapHost: imap.host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
        imapUser: imap.username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
        smtpHost: smtp.host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
        smtpUser: smtp.username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    )
}

/// accountwizard.BuildConfig: assembles the wire configuration from the
/// rows and their pins. It is the one place that sets the authentication
/// method.
public func buildConfig(identity id: Identity, name: String, imap: ServerFields, smtp: ServerFields) -> AccountConfig {
    let email = id.email.trimmingCharacters(in: .whitespacesAndNewlines)
    var accountName = name.trimmingCharacters(in: .whitespacesAndNewlines)
    if accountName.isEmpty {
        accountName = suggestAccountName(email)
    }
    return AccountConfig(
        name: accountName,
        email: email,
        displayName: nonEmpty(id.displayName.trimmingCharacters(in: .whitespacesAndNewlines)),
        kind: .imap,
        imap: serverConfig(imap),
        smtp: serverConfig(smtp)
    )
}

private func serverConfig(_ f: ServerFields) -> ServerConfig {
    ServerConfig(
        host: f.host.trimmingCharacters(in: .whitespacesAndNewlines),
        port: f.port,
        security: f.security,
        username: f.username.trimmingCharacters(in: .whitespacesAndNewlines),
        authMethod: .password,
        certificateSha256: nonEmpty(f.certificateSha256)
    )
}

/// accountwizard.credentialsFor: the secret of the identity page; an empty
/// password is no password (`account.update` then keeps the stored one).
public func credentialsFor(_ id: Identity) -> Credentials {
    Credentials(password: nonEmpty(id.password))
}

/// Go's "" sentinel as an Optional.
private func nonEmpty(_ s: String) -> String? {
    s.isEmpty ? nil : s
}
